Pagination: Fix page input bug.

timmyc opened the pull request:
While testing out #251 - I noticed that the page input field was not changing when I attempted to update the page manually. Using the arrows to toggle next/previous worked fine, but inputing a number or using the up/down arrows within the field did not update the page.

It seems there was a bit of a mish-mash of the page number being controlled from the parent component, and updated via `props.onPageChange`, and a value for page being tracked in component `state`. Since we want to persist report params in the URL, I opted to fix the bug here by moving all usage out of `state` and into using `props.onPageChange`.

__To Test__
- Open up `/wp-admin/admin.php?page=wc-admin#/analytics/revenue`
- Verify you can change pages by using the < > arrows, and the page number should update in the URL, the label next to the arrows, and the input field itself.
- Update the page number using the input field. Verify all is updated as expected.

timmyc committed:
Pagination: Fix page input bug.

psealock commented on client/components/pagination/index.js:
```diff
@@ -84,7 +62,8 @@ class Pagination extends Component {
 	onPageValueChange( event ) {
-		this.setState( { pageInputValue: event.target.value } );
+		const { onPageChange } = this.props;
+		onPageChange( parseInt( event.target.value ) );
```
Should use a [radix](https://davidwalsh.name/parseint-radix)

psealock commented on client/components/pagination/index.js:
```diff
@@ -26,27 +26,14 @@ function keyListener( methodToCall, event ) {
 		super( props );
-		this.state = {
-			pageInputValue: props.page,
-		};
```
Good refactor here. This is an anti-pattern, https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#anti-pattern-unconditionally-copying-props-to-state

psealock reviewed (commented):
Changes look good and code as well. Just one small comment. I also noticed that when you try to delete the last digit using the keyboard `NaN` becomes the query parameter.

![screen shot 2018-07-31 at 4 47 45 pm](https://user-images.githubusercontent.com/1922453/43438317-c158a242-94e1-11e8-8d74-ba1795614aa4.png)

![screen shot 2018-07-31 at 4 47 53 pm](https://user-images.githubusercontent.com/1922453/43438323-c509def6-94e1-11e8-9449-05d49f1b61cf.png)

timmyc committed:
Ensure NaN doesn't sneak in there.

timmyc commented:
Thanks for the review @psealock and for catching the NaN bug. Added some defensive logic around the input to ensure we have a valid number. I also added logic to select all in the input when it is clicked on to make using this feature a bit easier.

![go-to-page](https://user-images.githubusercontent.com/22080/43473725-cf352840-94a5-11e8-801b-b6aa4fb5d6ba.gif)

justinshreve commented on client/components/pagination/index.js:
```diff
@@ -84,7 +54,16 @@ class Pagination extends Component {
+		const { onPageChange } = this.props;
+		const newPage = parseInt( event.target.value, 10 );
+
+		if ( isFinite( newPage ) ) {
```
I think we still want to compare against `this.pageCount` before updating the URL -- otherwise this can happen:

<img width="722" alt="screen shot 2018-08-01 at 8 27 08 am" src="https://user-images.githubusercontent.com/689165/43521644-c670d730-9564-11e8-9647-900f78e7820c.png">

emdashcodes reviewed (commented):
Thanks for fixing this. I noticed one problem, but otherwise this is looking good to me.

timmyc committed:
Verify the page number exists.

timmyc commented:
Ah great catch @justinshreve ! Updated.

emdashcodes reviewed (approved):
👍
